### PR TITLE
Merge the providers into a single provider

### PR DESCRIPTION
**Pull request by steeling**
Part of [#4863](https://github.com/openservicemesh/osm/issues/4863)

Go favors composition, therefore there's no reason the providers need to be separated. That is, any implementations can simply combine multiple implementations through composition (although we deem it unlikely that a single provider would be sharded in the first place).

This is part of the layering of clients in the issue linked above, and gives us a canonical place to put methods that accept OSM object input, but need to query k8s.

Lastly, we embed the provider into the MeshCatalog, and remove its duplicative methods. If we ever need a mutli-provider, we can create one that fits the same interface.

Bonus points that this should make navigating code simpler since we're removing a layer of indirection

### CODE REVIEW

File: pkg/providers/kube/client.go [hunk 1]
```diff
@@ -15,8 +17,7 @@ import (
 )
 
 // Ensure interface compliance
-var _ endpoint.Provider = (*client)(nil)
-var _ service.Provider = (*client)(nil)
+var _ providers.Provider = (*client)(nil)
```
Comment (jsturtevant) [inline pkg/providers/kube/client.go:20]: til: https://go.dev/doc/faq#inheritance

**Review by jsturtevant — COMMENTED** [review]
like how this simplifies the usage.  Only comment is what type of interface name of provider is pretty generic?  What is it a provider of?  Is the idea in the future you could implement this interface for other components beyond Kubernetes?

File: pkg/endpoint/types.go [hunk 2]
```diff
@@ -7,27 +7,8 @@ package endpoint
...
-
-	// ListEndpointsForIdentity retrieves the list of IP addresses for the given service account
-	ListEndpointsForIdentity(identity.ServiceIdentity) []Endpoint
-
-	// GetResolvableEndpointsForService returns the expected endpoints that are to be reached when the service FQDN is resolved under
-	// the scope of the provider
-	GetResolvableEndpointsForService(service.MeshService) []Endpoint
-
-	// GetID returns the unique identifier of the EndpointsProvider.
-	GetID() string
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/endpoint/types.go:28]: > like how this simplifies the usage. Only comment is what type of interface name of provider is pretty generic? What is it a provider of? Is the idea in the future you could implement this interface for other components beyond Kubernetes?

Agreed the name is too generic. It is a provider/client of the underlying infra. I think I like `InfraClient` the most, but `ControlPlaneClient`, `DiscoveryClient` is also good. Curious on others thoughts here

[the remark this answers is not in the document]
Comment (shashankram) [reply pkg/endpoint/types.go:28]: I'd rename Provider to ComputeProvider

[the remark this answers is not in the document]
Comment (steeling) [reply pkg/endpoint/types.go:28]: Compute works! I'm getting super nitpicky here, but I'm personally not a fan of "Provider" to me that would indicate it is "Providing compute". "ComputeClient" indicates it is a "client to a compute service" in my head, so sounds better, and is a bit more common case. Or we could take a page of the k8s book and use `compute.Interface`

Not a hill I'm going to die on, just wanted to add as a counter

[the remark this answers is not in the document]
Comment (jsturtevant) [reply pkg/endpoint/types.go:28]: I like `Compute`.  Is this mostly for Network related things?  NetworkDiscovery?

[the remark this answers is not in the document]
Comment (steeling) [reply pkg/endpoint/types.go:28]: yup, but will eventually also include Policy definitions (retry policy, access policies, traffic splits, rate limiting).

**Review by keithmattix — APPROVED** [review]
Happy to see this change; even the idea of separate "endpoint" and "service" providers is a leaky abstraction that betrays Kubernetes design. LGTM

File: pkg/catalog/endpoint_test.go [hunk 3]
```diff
@@ -139,22 +140,20 @@ func TestListAllowedUpstreamEndpointsForService(t *testing.T) {
 
 			mockConfigurator := configurator.NewMockConfigurator(mockCtrl)
 			mockKubeController := k8s.NewMockController(mockCtrl)
-			mockEndpointProvider := endpoint.NewMockProvider(mockCtrl)
-			mockServiceProvider := service.NewMockProvider(mockCtrl)
+			mockProvider := compute.NewMockInterface(mockCtrl)
```
Comment (jaellio) [inline pkg/catalog/endpoint_test.go:143]: nit: This applies to a couple of other files as well
```suggestion
			mockComputeInterface := compute.NewMockInterface(mockCtrl)
```

**Review by jaellio — APPROVED** [review]
LGTM

**Comment by steeling** [comment]
Unrlated lint errors, going to merge

**Pull request merged**